What is the correct way for running update-codegen and go test?

I'm having trouble identifying the correct way to run the `update-codegen.sh` and `go test ./...`

I'm working on a project that also has kubernetes controllers and it has go modules setup within it. It's very similar to this sample-controller project. We are looking for a good way to get the `k8s.io/code-generator` dependency in the `vendor` directory. Although it seems that this sample-controller project doesn't have a vendor directory in the first place.
 
**Q1:** Is it intended for us to run `go mod vendor` to get the codegen dependencies in the correct place?
**Q2:** Are we expected to use the **latest version** of `k8s.io/code-generator`? That is, I'm having a hard time pinning the version of `k8s.io/code-generator`.

**Environment**
```
$ go version
go version go1.12.4 darwin/amd64
$ go env
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/wfernandes/Library/Caches/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/wfernandes/go"
GOPROXY=""
GORACE=""
GOROOT="/usr/local/go"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/w9/_qjbbv15667cxfs7vpply4bh0000gn/T/go-build133647668=/tmp/go-build -gno-record-gcc-switches -fno-common"
```
**Steps to Reproduce**

1. Ensure your GOPATH is set correctly
1. `go get github.com/kubernetes/sample-controller`
   I get the following output when I do the above `go get`
   ```
   $ go get github.com/kubernetes/sample-controller
   # k8s.io/sample-controller/pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1
   ../../k8s.io/sample-controller/pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1/samplecontroller _client.go:73:45: scheme.Codecs.WithoutConversion undefined (type serializer.CodecFactory has no field or method WithoutConversion)
   ```
1. `cd $GOPATH/src/github.com/kubernetes/sample-controller`
1. `./hack/update-codegen`
     I get the following output
     ```
     ./hack/update-codegen.sh: line 28: ../code-generator/generate-groups.sh: No such file or directory
     ```
1. Upon inspection of the script, I noticed that it is referring to the `vendor` directory.
1. `GO111MODULE=on go mod vendor`
     This creates a vendor directory in the project with the `k8s.io/code-generator` dependency.
1. `./hack/update-codegen.sh`
    I get the following output.
    ```
     ./hack/update-codegen.sh: line 28: ./vendor/k8s.io/code-generator/generate-groups.sh: Permission denied
    ```
1. I edited the `hack/update-codegen.sh` script by updating the command to the following
   ```
    bash "${CODEGEN_PKG}"/generate-groups.sh "deepcopy,client,informer,lister" ....
   ```
1. `./hack/update-codegen.sh`
   This gets update-codegen to work!!
   ```
   Generating deepcopy funcs
   Generating clientset for samplecontroller:v1alpha1 at k8s.io/sample-controller/pkg/generated/clientset
   Generating listers for samplecontroller:v1alpha1 at k8s.io/sample-controller/pkg/generated/listers
   Generating informers for samplecontroller:v1alpha1 at k8s.io/sample-controller/pkg/generated/informers
   ```
1. But now when I run `go test ./...`
```
# k8s.io/sample-controller/pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1
../../../k8s.io/sample-controller/pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1/samplecontroller_client.go:73:45: scheme.Codecs.WithoutConversion undefined (type serializer.CodecFactory has no field or method WithoutConversion)
# github.com/kubernetes/sample-controller/pkg/generated/clientset/versioned/scheme
pkg/generated/clientset/versioned/scheme/register.go:34:26: cannot use v1alpha1.AddToScheme (type func(*"k8s.io/apimachinery/pkg/runtime".Scheme) error) as type func(*"github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Scheme) error in array or slice literal
# github.com/kubernetes/sample-controller/pkg/generated/listers/samplecontroller/v1alpha1
pkg/generated/listers/samplecontroller/v1alpha1/foo.go:91:51: cannot use v1alpha1.Resource("foo") (type "k8s.io/apimachinery/pkg/runtime/schema".GroupResource) as type "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime/schema".GroupResource in argument to "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/api/errors".NewNotFound
FAIL	github.com/kubernetes/sample-controller [build failed]
?   	github.com/kubernetes/sample-controller/pkg/apis/samplecontroller	[no test files]
# github.com/kubernetes/sample-controller/pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1
pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1/foo.go:73:19: cannot use scheme.ParameterCodec (type "k8s.io/apimachinery/pkg/runtime".ParameterCodec) as type "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".ParameterCodec in argument to c.client.Get().Namespace(c.ns).Resource("foos").Name(name).VersionedParams:
	"k8s.io/apimachinery/pkg/runtime".ParameterCodec does not implement "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".ParameterCodec (wrong type for DecodeParameters method)
		have DecodeParameters(url.Values, "k8s.io/apimachinery/pkg/runtime/schema".GroupVersion, "k8s.io/apimachinery/pkg/runtime".Object) error
		want DecodeParameters(url.Values, "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime/schema".GroupVersion, "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object) error
pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1/foo.go:75:7: cannot use result (type *v1alpha1.Foo) as type "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object in argument to c.client.Get().Namespace(c.ns).Resource("foos").Name(name).VersionedParams(&options, scheme.ParameterCodec).Do().Into:
	*v1alpha1.Foo does not implement "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object (wrong type for DeepCopyObject method)
		have DeepCopyObject() "k8s.io/apimachinery/pkg/runtime".Object
		want DeepCopyObject() "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object
pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1/foo.go:89:19: cannot use scheme.ParameterCodec (type "k8s.io/apimachinery/pkg/runtime".ParameterCodec) as type "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".ParameterCodec in argument to c.client.Get().Namespace(c.ns).Resource("foos").VersionedParams:
	"k8s.io/apimachinery/pkg/runtime".ParameterCodec does not implement "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".ParameterCodec (wrong type for DecodeParameters method)
		have DecodeParameters(url.Values, "k8s.io/apimachinery/pkg/runtime/schema".GroupVersion, "k8s.io/apimachinery/pkg/runtime".Object) error
		want DecodeParameters(url.Values, "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime/schema".GroupVersion, "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object) error
pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1/foo.go:92:7: cannot use result (type *v1alpha1.FooList) as type "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object in argument to c.client.Get().Namespace(c.ns).Resource("foos").VersionedParams(&opts, scheme.ParameterCodec).Timeout(timeout).Do().Into:
	*v1alpha1.FooList does not implement "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object (wrong type for DeepCopyObject method)
		have DeepCopyObject() "k8s.io/apimachinery/pkg/runtime".Object
		want DeepCopyObject() "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object
pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1/foo.go:106:19: cannot use scheme.ParameterCodec (type "k8s.io/apimachinery/pkg/runtime".ParameterCodec) as type "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".ParameterCodec in argument to c.client.Get().Namespace(c.ns).Resource("foos").VersionedParams:
	"k8s.io/apimachinery/pkg/runtime".ParameterCodec does not implement "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".ParameterCodec (wrong type for DecodeParameters method)
		have DecodeParameters(url.Values, "k8s.io/apimachinery/pkg/runtime/schema".GroupVersion, "k8s.io/apimachinery/pkg/runtime".Object) error
		want DecodeParameters(url.Values, "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime/schema".GroupVersion, "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object) error
pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1/foo.go:119:7: cannot use result (type *v1alpha1.Foo) as type "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object in argument to c.client.Post().Namespace(c.ns).Resource("foos").Body(foo).Do().Into:
	*v1alpha1.Foo does not implement "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object (wrong type for DeepCopyObject method)
		have DeepCopyObject() "k8s.io/apimachinery/pkg/runtime".Object
		want DeepCopyObject() "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object
pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1/foo.go:132:7: cannot use result (type *v1alpha1.Foo) as type "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object in argument to c.client.Put().Namespace(c.ns).Resource("foos").Name(foo.ObjectMeta.Name).Body(foo).Do().Into:
	*v1alpha1.Foo does not implement "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object (wrong type for DeepCopyObject method)
		have DeepCopyObject() "k8s.io/apimachinery/pkg/runtime".Object
		want DeepCopyObject() "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object
pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1/foo.go:148:7: cannot use result (type *v1alpha1.Foo) as type "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object in argument to c.client.Put().Namespace(c.ns).Resource("foos").Name(foo.ObjectMeta.Name).SubResource("status").Body(foo).Do().Into:
	*v1alpha1.Foo does not implement "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object (wrong type for DeepCopyObject method)
		have DeepCopyObject() "k8s.io/apimachinery/pkg/runtime".Object
		want DeepCopyObject() "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object
pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1/foo.go:172:19: cannot use scheme.ParameterCodec (type "k8s.io/apimachinery/pkg/runtime".ParameterCodec) as type "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".ParameterCodec in argument to c.client.Delete().Namespace(c.ns).Resource("foos").VersionedParams:
	"k8s.io/apimachinery/pkg/runtime".ParameterCodec does not implement "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".ParameterCodec (wrong type for DecodeParameters method)
		have DecodeParameters(url.Values, "k8s.io/apimachinery/pkg/runtime/schema".GroupVersion, "k8s.io/apimachinery/pkg/runtime".Object) error
		want DecodeParameters(url.Values, "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime/schema".GroupVersion, "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object) error
pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1/foo.go:189:7: cannot use result (type *v1alpha1.Foo) as type "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object in argument to c.client.Patch(pt).Namespace(c.ns).Resource("foos").SubResource(subresources...).Name(name).Body(data).Do().Into:
	*v1alpha1.Foo does not implement "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object (wrong type for DeepCopyObject method)
		have DeepCopyObject() "k8s.io/apimachinery/pkg/runtime".Object
		want DeepCopyObject() "github.com/kubernetes/sample-controller/vendor/k8s.io/apimachinery/pkg/runtime".Object
pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1/foo.go:189:7: too many errors
?   	github.com/kubernetes/sample-controller/pkg/apis/samplecontroller/v1alpha1	[no test files]
```

--- a/staging/src/k8s.io/apiextensions-apiserver/examples/client-go/hack/update-codegen.sh
+++ b/staging/src/k8s.io/apiextensions-apiserver/examples/client-go/hack/update-codegen.sh
@@ -25,7 +25,7 @@ CODEGEN_PKG=${CODEGEN_PKG:-$(cd "${SCRIPT_ROOT}"; ls -d -1 ./vendor/k8s.io/code-
 # --output-base    because this script should also be able to run inside the vendor dir of
 #                  k8s.io/kubernetes. The output-base is needed for the generators to output into the vendor dir
 #                  instead of the $GOPATH directly. For normal projects this can be dropped.
-"${CODEGEN_PKG}/generate-groups.sh" all \
+bash "${CODEGEN_PKG}/generate-groups.sh" all \
   k8s.io/apiextensions-apiserver/examples/client-go/pkg/client k8s.io/apiextensions-apiserver/examples/client-go/pkg/apis \
   cr:v1 \
   --output-base "$(dirname "${BASH_SOURCE[0]}")/../../../../.."

--- a/staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh
+++ b/staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh
@@ -27,7 +27,7 @@ CODEGEN_PKG=${CODEGEN_PKG:-$(cd "${SCRIPT_ROOT}"; ls -d -1 ./vendor/k8s.io/code-
 #                  instead of the $GOPATH directly. For normal projects this can be dropped.
 CLIENTSET_NAME_VERSIONED=clientset \
 CLIENTSET_NAME_INTERNAL=internalclientset \
-"${CODEGEN_PKG}/generate-internal-groups.sh" all \
+bash "${CODEGEN_PKG}/generate-internal-groups.sh" all \
   k8s.io/apiextensions-apiserver/pkg/client k8s.io/apiextensions-apiserver/pkg/apis k8s.io/apiextensions-apiserver/pkg/apis \
   "apiextensions:v1beta1" \
   --output-base "$(dirname "${BASH_SOURCE[0]}")/../../.." \

--- a/staging/src/k8s.io/kube-aggregator/hack/update-codegen.sh
+++ b/staging/src/k8s.io/kube-aggregator/hack/update-codegen.sh
@@ -23,7 +23,7 @@ CODEGEN_PKG=${CODEGEN_PKG:-$(cd "${SCRIPT_ROOT}"; ls -d -1 ./vendor/k8s.io/code-
 
 CLIENTSET_NAME_VERSIONED=clientset \
 CLIENTSET_PKG_NAME=clientset_generated \
-"${CODEGEN_PKG}/generate-groups.sh" deepcopy,client,lister,informer \
+bash "${CODEGEN_PKG}/generate-groups.sh" deepcopy,client,lister,informer \
   k8s.io/kube-aggregator/pkg/client k8s.io/kube-aggregator/pkg/apis \
   "apiregistration:v1beta1,v1" \
   --output-base "$(dirname "${BASH_SOURCE[0]}")/../../.." \
@@ -32,7 +32,7 @@ CLIENTSET_PKG_NAME=clientset_generated \
 CLIENTSET_NAME_VERSIONED=clientset \
 CLIENTSET_PKG_NAME=clientset_generated \
 CLIENTSET_NAME_INTERNAL=internalclientset \
-"${CODEGEN_PKG}/generate-internal-groups.sh" deepcopy,client,lister,informer,conversion \
+bash "${CODEGEN_PKG}/generate-internal-groups.sh" deepcopy,client,lister,informer,conversion \
   k8s.io/kube-aggregator/pkg/client k8s.io/kube-aggregator/pkg/apis k8s.io/kube-aggregator/pkg/apis \
   "apiregistration:v1beta1,v1" \
   --output-base "$(dirname "${BASH_SOURCE[0]}")/../../.." \

--- a/staging/src/k8s.io/metrics/hack/update-codegen.sh
+++ b/staging/src/k8s.io/metrics/hack/update-codegen.sh
@@ -28,12 +28,12 @@ CODEGEN_PKG=${CODEGEN_PKG:-$(cd "${SCRIPT_ROOT}"; ls -d -1 ./vendor/k8s.io/code-
 #
 # we skip informers and listers for metrics, because we don't quite support the requisite operations yet
 # we skip generating the internal clientset as it's not really needed
-"${CODEGEN_PKG}/generate-internal-groups.sh" deepcopy,conversion \
+bash "${CODEGEN_PKG}/generate-internal-groups.sh" deepcopy,conversion \
   k8s.io/metrics/pkg/client k8s.io/metrics/pkg/apis k8s.io/metrics/pkg/apis \
   "metrics:v1alpha1,v1beta1 custom_metrics:v1beta1 external_metrics:v1beta1" \
   --output-base "$(dirname "${BASH_SOURCE[0]}")/../../.." \
   --go-header-file "${SCRIPT_ROOT}/hack/boilerplate.go.txt"
-"${CODEGEN_PKG}/generate-groups.sh" client \
+bash "${CODEGEN_PKG}/generate-groups.sh" client \
   k8s.io/metrics/pkg/client k8s.io/metrics/pkg/apis \
   "metrics:v1alpha1,v1beta1" \
   --output-base "$(dirname "${BASH_SOURCE[0]}")/../../.." \

--- a/staging/src/k8s.io/node-api/hack/update-codegen.sh
+++ b/staging/src/k8s.io/node-api/hack/update-codegen.sh
@@ -25,7 +25,7 @@ CODEGEN_PKG=${CODEGEN_PKG:-$(cd "${SCRIPT_ROOT}"; ls -d -1 ./vendor/k8s.io/code-
 # --output-base    because this script should also be able to run inside the vendor dir of
 #                  k8s.io/kubernetes. The output-base is needed for the generators to output into the vendor dir
 #                  instead of the $GOPATH directly. For normal projects this can be dropped.
-"${CODEGEN_PKG}/generate-groups.sh" "deepcopy,client,informer,lister" \
+bash "${CODEGEN_PKG}/generate-groups.sh" "deepcopy,client,informer,lister" \
   k8s.io/node-api/pkg/client k8s.io/node-api/pkg/apis \
   "node:v1alpha1" \
   --output-base "$(dirname "${BASH_SOURCE[0]}")/../../.." \

--- a/staging/src/k8s.io/sample-apiserver/README.md
+++ b/staging/src/k8s.io/sample-apiserver/README.md
@@ -55,7 +55,7 @@ Note, however, that if you intend to
 [generate code](#changes-to-the-types) then you will also need the
 code-generator repo to exist in an old-style location.  One easy way
 to do this is to use the command `go mod vendor` to create and
-popdulate the `vendor` directory.
+populate the `vendor` directory.
 
 ### A Note on kubernetes/kubernetes
 

--- a/staging/src/k8s.io/sample-controller/README.md
+++ b/staging/src/k8s.io/sample-controller/README.md
@@ -31,6 +31,47 @@ The sample controller uses [client-go library](https://github.com/kubernetes/cli
 The details of interaction points of the sample controller with various mechanisms from this library are
 explained [here](docs/controller-client-go.md).
 
+## Fetch sample-controller and its dependencies
+
+Like the rest of Kubernetes, sample-controller has used
+[godep](https://github.com/tools/godep) and `$GOPATH` for years and is
+now adopting go 1.11 modules.  There are thus two alternative ways to
+go about fetching this demo and its dependencies.
+
+### Fetch with godep
+
+When NOT using go 1.11 modules, you can use the following commands.
+
+```sh
+go get -d k8s.io/sample-controller
+cd $GOPATH/src/k8s.io/sample-controller
+godep restore
+```
+
+### When using go 1.11 modules
+
+When using go 1.11 modules (`GO111MODULE=on`), issue the following
+commands --- starting in whatever working directory you like.
+
+```sh
+git clone https://github.com/kubernetes/sample-controller.git
+cd sample-controller
+```
+
+Note, however, that if you intend to
+[generate code](#changes-to-the-types) then you will also need the
+code-generator repo to exist in an old-style location.  One easy way
+to do this is to use the command `go mod vendor` to create and
+populate the `vendor` directory.
+
+### A Note on kubernetes/kubernetes
+
+If you are developing Kubernetes according to
+https://github.com/kubernetes/community/blob/master/contributors/guide/github-workflow.md
+then you already have a copy of this demo in
+`kubernetes/staging/src/k8s.io/sample-controller` and its dependencies
+--- including the code generator --- are in usable locations
+(valid for all Go versions).
 
 ## Purpose
 
@@ -42,8 +83,6 @@ This is an example of how to build a kube-like controller with a single type.
 
 ```sh
 # assumes you have a working kubeconfig, not required if operating in-cluster
-go get k8s.io/sample-controller
-cd $GOPATH/src/k8s.io/sample-controller
 go build -o sample-controller .
 ./sample-controller -kubeconfig=$HOME/.kube/config
 

--- a/staging/src/k8s.io/sample-controller/hack/update-codegen.sh
+++ b/staging/src/k8s.io/sample-controller/hack/update-codegen.sh
@@ -25,7 +25,7 @@ CODEGEN_PKG=${CODEGEN_PKG:-$(cd "${SCRIPT_ROOT}"; ls -d -1 ./vendor/k8s.io/code-
 # --output-base    because this script should also be able to run inside the vendor dir of
 #                  k8s.io/kubernetes. The output-base is needed for the generators to output into the vendor dir
 #                  instead of the $GOPATH directly. For normal projects this can be dropped.
-"${CODEGEN_PKG}"/generate-groups.sh "deepcopy,client,informer,lister" \
+bash "${CODEGEN_PKG}"/generate-groups.sh "deepcopy,client,informer,lister" \
   k8s.io/sample-controller/pkg/generated k8s.io/sample-controller/pkg/apis \
   samplecontroller:v1alpha1 \
   --output-base "$(dirname "${BASH_SOURCE[0]}")/../../.." \

